src/theory/strings/theory_strings_rewriter.cpp

```diff
@@ -2219,6 +2219,42 @@ Node TheoryStringsRewriter::rewriteReplace( Node node ) {
     }
   }
 
+  children1.clear();
+  getConcat(node[1], children1);
+  Node lastChild1 = children1[children1.size() - 1];
+  if (lastChild1.getKind() == kind::STRING_SUBSTR)
+  {
+    // (str.replace x (str.++ ... (str.substr y i j)) z) --->
```

Comment:
If you replace each occurrence of "..." with e.g. "t" this would be clearer I think.